Alex endpoint templates

alek-thunder opened the pull request.

alek-thunder committed:
Added possibility to define endpoints as templates in order to randomly select from a range of numbered video files

alek-thunder committed:
Adjusted REST API POST and PUT handling of endpoints

alek-thunder committed:
Adjusted variable naming

antongulenko commented on rest-api.go:
```diff
@@ -25,23 +25,35 @@ func (api *SetUrlsRestApi) handleEndpoints(writer http.ResponseWriter, req *http
 		if len(lines) > 0 {
-			api.Col.Factory.URLs = lines
+			api.Col.Factory.hosts = nil
+			api.Col.Factory.hostURLs = nil
```
initialize the map with make()

antongulenko commented on rest-api.go:
```diff
@@ -25,23 +25,35 @@ func (api *SetUrlsRestApi) handleEndpoints(writer http.ResponseWriter, req *http
+					api.Col.Factory.hosts = append(api.Col.Factory.hosts, host)
+					api.Col.Factory.hostURLs[host] = append(api.Col.Factory.hostURLs[host], urls...)
+				} else {
+					writer.Write([]byte(fmt.Sprintf("Error handling streaming endpoint %v. Reason: %v", entry, er)))
```
Also log the error

antongulenko commented on rest-api.go:
```diff
@@ -25,23 +25,35 @@ func (api *SetUrlsRestApi) handleEndpoints(writer http.ResponseWriter, req *http
+
 		if len(lines) > 0 {
-			api.Col.Factory.URLs = append(api.Col.Factory.URLs, lines...)
+			for _, entry := range lines {
```
This block seems duplicated from POST - move to extra method

antongulenko commented on rest-api.go:
```diff
@@ -25,23 +25,35 @@ func (api *SetUrlsRestApi) handleEndpoints(writer http.ResponseWriter, req *http
 	}
-	urls := api.Col.Factory.URLs
-	writer.Write([]byte(fmt.Sprintf("%v active endpoint(s):\n", len(urls))))
-	for _, endpoint := range urls {
```
Why not reply the list of endpoints when everything was successfull?

alek-thunder replied:
is done in appendEndpointURLs method further below

antongulenko commented on rtmp-stream.go:
```diff
@@ -18,32 +22,44 @@ const maxRtmpChannelNumber = 100
-	urls := f.URLs
-	if len(urls) == 0 {
+func (f *RtmpStreamFactory) nextURL() (*url.URL, error) {
+	rand.Seed(time.Now().Unix())
```
The RNG should not be re-seeded every time... Initialize only once.

antongulenko commented on rtmp-stream.go:
```diff
@@ -18,32 +22,44 @@ const maxRtmpChannelNumber = 100
+func (f *RtmpStreamFactory) nextURL() (*url.URL, error) {
+	rand.Seed(time.Now().Unix())
+	for i := len(f.hosts); i >= 0; i-- {
+		if nextHost, er := f.nextHost(); er != nil {
```
error variables are called 'err' by convention.
Also, the error is discarded here - at least log it.

antongulenko commented on rtmp-stream.go:
```diff
@@ -18,32 +22,44 @@ const maxRtmpChannelNumber = 100
+		if nextHost, er := f.nextHost(); er != nil {
+			return nil, ErrorNoURLs
+		} else {
+			if len(f.hostURLs[nextHost]) > 0 { // Success
```
Make more readable: extract f.hostURLs[nextHost] into a variable, and also the rand.Intn() call.

antongulenko commented on rtmp-stream.go:
```diff
@@ -98,6 +114,67 @@ func (f *RtmpStreamFactory) startStream(conn rtmp.ClientConn, streamName string)
+func (f *RtmpStreamFactory) ParseURLArgument(urlArg string) (string, []*url.URL, error) {
+	var host string
+	var urls []*url.URL
+	var r = regexp.MustCompile(`{{(?P<min>[1-9][0-9]*) (?P<max>[1-9][0-9]*)}}`) // {{123 456}}
```
Excape curly braces? Not sure.

antongulenko commented on rtmp-stream.go:
```diff
@@ -98,6 +114,67 @@ func (f *RtmpStreamFactory) startStream(conn rtmp.ClientConn, streamName string)
+	var urls []*url.URL
+	var r = regexp.MustCompile(`{{(?P<min>[1-9][0-9]*) (?P<max>[1-9][0-9]*)}}`) // {{123 456}}
+
+	var regexString = "'{{(?P<min>[1-9][0-9]*) (?P<max>[1-9][0-9]*)}}'"
```
Duplicate string constant: use the regexString variable as input to regexp.MustCompile()

antongulenko commented on rtmp-stream.go:
```diff
@@ -98,6 +114,67 @@ func (f *RtmpStreamFactory) startStream(conn rtmp.ClientConn, streamName string)
+	var r = regexp.MustCompile(`{{(?P<min>[1-9][0-9]*) (?P<max>[1-9][0-9]*)}}`) // {{123 456}}
+
+	var regexString = "'{{(?P<min>[1-9][0-9]*) (?P<max>[1-9][0-9]*)}}'"
+	var regexInfo = fmt.Sprintf("Use regex that matches patter %v", regexString)
```
patterN

antongulenko commented on rtmp-stream.go:
```diff
@@ -98,6 +114,67 @@ func (f *RtmpStreamFactory) startStream(conn rtmp.ClientConn, streamName string)
+	var regexInfo = fmt.Sprintf("Use regex that matches patter %v", regexString)
+
+	if r.MatchString(urlArg) { // URL is a template.
+		log.Infoln("Processing template URL %v with regex matching. (Used regex: '%v')", urlArg, regexString)
```
Use log.Infof instead of Infoln when using formatting sequences (also below, check all log calls).

antongulenko commented on rtmp-stream.go:
```diff
@@ -98,6 +114,67 @@ func (f *RtmpStreamFactory) startStream(conn rtmp.ClientConn, streamName string)
+
+	if r.MatchString(urlArg) { // URL is a template.
+		log.Infoln("Processing template URL %v with regex matching. (Used regex: '%v')", urlArg, regexString)
+		match := r.FindStringSubmatch(urlArg)
```
The regex is matched a second time here. Instead of using MatchString() and FindStringSubmatch separately, ues the result of FindStringSubmatch to find out IF there was a match.

alek-thunder replied:
All other comments are fixed. From performance perspective I  agree with this one. However, it is not function that  is called often and the way it is now is better readable from my point of view.

antongulenko commented on rtmp-stream.go:
```diff
@@ -98,6 +114,67 @@ func (f *RtmpStreamFactory) startStream(conn rtmp.ClientConn, streamName string)
+			if err != nil {
+				return "", nil, fmt.Errorf("Failed to parse URL (%v): %v", unparsedURL, err)
+			}
+			host= parsedURL.Host
```
format code ;)

alek-thunder replied:
?

antongulenko replied:
Space after the variable name... This should be handled automatically by gofmt.

antongulenko commented on rtmp-stream.go:
```diff
@@ -98,6 +114,67 @@ func (f *RtmpStreamFactory) startStream(conn rtmp.ClientConn, streamName string)
+	if strings.Contains(urlArg, toReplace) {
+		for i := min; i <= max; i++ {
+			unparsedURL := strings.Replace(urlArg, toReplace, strconv.Itoa(i), 1)
+			parsedURL, err := url.Parse(unparsedURL)
```
URL parsing is duplicated here and in the else-part of ParseURLArgument(). Suggestion: First generate all URLs (in case there is no template, the "generation" is trivial), THEN do the URL parsing on all resulting URLs.

alek-thunder committed:
Fixed code duplicate for stream endpoint template url parsing. Fixed some other minor stuff.

alek-thunder committed:
Allows to test whether it is possible to connect to all initially defined endpoints. Result is logged.

antongulenko commented on rtmp-stream.go:
```diff
@@ -98,6 +113,78 @@ func (f *RtmpStreamFactory) startStream(conn rtmp.ClientConn, streamName string)
+		if min > max {
+			return "", nil, fmt.Errorf("Minimal value cannot be greater than maximal value in url %v. %v.", urlArg, regexInfo)
+		}
+		if urls, er := f.generateURLs(urlArg, match[0], min, max); er == nil {
```
Call the error variable "err"

alek-thunder replied:
should be replaced everywhere

antongulenko commented on rtmp-stream.go:
```diff
@@ -98,6 +113,78 @@ func (f *RtmpStreamFactory) startStream(conn rtmp.ClientConn, streamName string)
+	for _, unparsedURL := range unparsedURLs {
+		parsedURL, err := url.Parse(unparsedURL)
+		if err != nil {
+			if len(errorMsg) == 0 {
```
For the issue of collecting multiple errors, there is the type golib.MultiError. Declare value and use Add() method, afterwards use NilOrError() to get a combined error (or nil, if there is no error).
I suggest returning the resulting error immediately after the loop, if it is non-nil.

alek-thunder committed:
- wrote distribution argument parser tests
 - several bugfixes
 - cleaned code
 - fixed error variable naming to match convention
 - moved regex string as global const

antongulenko commented on main.go:
```diff
@@ -22,31 +23,45 @@ func main() {
+		if *testEndpoints {
+			summary, err := factory.TestAllEndpointURLs()
+			if err == nil {
+				log.Info(summary)
```
The Info() line is duplicated and can be pulled before the if clause

antongulenko commented on random-distribution-helpers.go:
```diff
@@ -14,60 +14,66 @@ type RandomDistribution interface {
-var _ RandomDistribution = RandomConstDistribution{}
+var _ RandomDistribution = &RandomConstDistribution{}
 
 type RandomConstDistribution struct {
```
Since this always returns the same value, shouldn't it just be called ConstDistribution?
IMO, the interface should probably also be called Distribution instead of RandomDistribution? But even if that's not the case, the constant one can be called ConstDistribution, even though it implements an interface that starts with "Random".

alek-thunder replied:
Distribution can be random while outputting one value. It is a distribution that has the probability of 1 for this value. RandomConstDistribution is a valid name

antongulenko replied:
Ok, but is the "random" really necessary? Isn't the statistical term just "distribution"? At least the titles on Wikipedia are all just "distribution" and "Verteilung", e.g. "continuous distribution", not "continuous random distribution".

alek-thunder replied:
Does not matter. Definitions alter (e.g. http://mathworld.wolfram.com/RandomDistribution.html ) but will adjust in order to shorten the naming

antongulenko commented on random-distribution-helpers.go:
```diff
@@ -14,60 +14,66 @@ type RandomDistribution interface {
-func (constDist RandomConstDistribution) String() string {
-	return fmt.Sprintf("Constant value: %d.", constDist.value)
+func (constDist *RandomConstDistribution) String() string {
+	return fmt.Sprintf("Constant value: %v ms.", int64(constDist.value/time.Millisecond))
```
For printing time.Duration values, just use %v (without dividing the value, or adding any "ms" suffix yourself). In general: Almost always just use %v when printing anything.

alek-thunder replied:
It prints the value as nanoseconds which looks ugly. I'd like to have it in milliseconds

antongulenko replied:
fmt.Println(3 * time.Minute + 3 * time.Second + 3 * time.Millisecond)
prints:
"3m3.003s"

You probably used "%d" as format character, which prints it as a number, without using the String() method defined on the time.Duration type.

alek-thunder replied:
Yes, done

antongulenko commented on random-distribution-helpers.go:
```diff
@@ -14,60 +14,66 @@ type RandomDistribution interface {
-	return fmt.Sprintf("Equal distribution between %d and %d.", equalDist.min, equalDist.max)
+func (equalDist *RandomEqualDistribution) String() string {
+	return fmt.Sprintf("Equal distribution between %v ms and %v ms.",
+		int64(equalDist.min/time.Millisecond), int64(equalDist.max/time.Millisecond))
```
See above comment about printing time.Duration values.

antongulenko replied:
Also check the other places where you do that.

alek-thunder replied:
See comment above

antongulenko replied:
See reply above

antongulenko commented on random-distribution-helpers.go:
```diff
@@ -82,43 +88,53 @@ func (distSampler RandomDistributionSampler) Set(value string) error {
+	} else if err == nil && timeValue < 0 {
+		return -1, fmt.Errorf("Distribution argument must be a positive time value but actually is %v", timeValue)
+	}
+	return -1, err
```
The if conditions in this method can be formulated in a way that the -1 return value (which cannot be reached) is not necessary. First check for the error, and return if it is non-nil, then check if timeValue is above or below zero.
Also, Since this method does not use the receiver distSampler, it should be turned into a function.

alek-thunder replied:
I re-adjusted the order of return values. However, both -1 values can be reached. First, the format can be invalid, thus the parsing fails. Second, the time value must not be negative (the Duration-parsing allows negative values). This requires a different error message imo. Third, everything is fine.

antongulenko commented on rtmp-stream.go:
```diff
@@ -31,11 +36,13 @@ type RtmpStreamFactory struct {
-				return f.hostURLs[nextHost][rand.Intn(len(f.hostURLs[nextHost]))], nil // Pick random URL of that host
+				hostLen := len(f.hostURLs[nextHost])
+				randomIndex := rand.Intn(hostLen)
+				return f.hostURLs[nextHost][rand.Intn(randomIndex)], nil // Pick random URL of that host
```
I think the second rand.Intn() is not necessary...?

antongulenko commented on rtmp-stream.go:
```diff
@@ -117,58 +158,50 @@ func (f *RtmpStreamFactory) ParseURLArgument(urlArg string) (string, []*url.URL,
-		return host, urls, nil
-	} else {
-		return "", nil, fmt.Errorf("Following urls that were generated from template url %v could not be parsed: %v", errorMsg)
+		return "", nil, fmt.Errorf("Failed to parse streaming endpoint urls from template %v: ", urlArg, err)
```
Errorf() has one format template, but gets two parmeters

alek-thunder committed:
Minor fixes

alek-thunder committed:
- renamed distribution structs, removed "Random" prefix
- some minor fixes
- some minor bugfixes

alek-thunder committed:
String format fix
